samza-core/src/main/java/org/apache/samza/runtime/ContainerLaunchUtil.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.runtime;
+
+import org.apache.samza.SamzaException;
+import org.apache.samza.application.descriptors.ApplicationDescriptor;
+import org.apache.samza.application.descriptors.ApplicationDescriptorImpl;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.ShellCommandConfig;
+import org.apache.samza.container.ContainerHeartbeatClient;
+import org.apache.samza.container.ContainerHeartbeatMonitor;
+import org.apache.samza.container.LocalityManager;
+import org.apache.samza.container.SamzaContainer;
+import org.apache.samza.container.SamzaContainer$;
+import org.apache.samza.container.SamzaContainerListener;
+import org.apache.samza.context.ExternalContext;
+import org.apache.samza.context.JobContextImpl;
+import org.apache.samza.job.model.JobModel;
+import org.apache.samza.metrics.MetricsRegistryMap;
+import org.apache.samza.metrics.MetricsReporter;
+import org.apache.samza.task.TaskFactory;
+import org.apache.samza.task.TaskFactoryUtil;
+import org.apache.samza.util.ScalaJavaUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Option;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class ContainerLaunchUtil {
+  private static final Logger log = LoggerFactory.getLogger(ContainerLaunchUtil.class);
+
+  private static volatile Throwable containerRunnerException = null;
+
+  /**
+   * This method launches a Samza container in a managed cluster, e.g. Yarn.
+   *
+   * NOTE: this util method is also invoked by Beam SamzaRunner.
+   * Any change here needs to take Beam into account.
+   */
+  public static void run(
+      ApplicationDescriptorImpl<? extends ApplicationDescriptor> appDesc,
+      String containerId,
+      JobModel jobModel) {
+
+    Config config = jobModel.getConfig();
+    run(appDesc, containerId, jobModel, config, buildExternalContext(config));
+
+    System.exit(0);
```

Comment:
both exit() and the below heart beat monitor are needed in the samza localContainerRUnner as well as beam container runner. So I don't see any reason to move those two logic out. Otherwise I have to duplicate the logic in beam container runner which seems not a good code reuse.

Generally there is no different in running the container in samza or beam. The only difference is the entry point, where Beam starts from user main() and has to use the application runner to run it, while Samza runs directly from LocalApplicationRunner.main().